api/admin/controller.py

```diff
@@ -631,3 +631,61 @@ def circulation_events(self):
         }, results)
 
         return dict({ "circulation_events": events })
+
+    def bulk_circulation_events(self):
+        annotator = AdminAnnotator(self.circulation)
+
+        start = flask.request.args.get("start")
+        end = flask.request.args.get("end")
+
+        from sqlalchemy.sql import func
+
+        t = self._db.query(
+            WorkGenre.work_id,
+            func.max(WorkGenre.affinity).label('max_affinity'),
+        ).group_by(WorkGenre.work_id).subquery()
+
+        query = self._db.query(
+                CirculationEvent, Identifier, Work, Edition, Genre
+            ) \
+            .join(LicensePool) \
+            .join(Work) \
+            .join(LicensePool.presentation_edition) \
+            .join(Edition.primary_identifier) \
+            .join(Work.work_genres) \
+            .join(WorkGenre.genre) \
+            .join(t, t.c.work_id == Work.id) \
+            .filter(WorkGenre.affinity == t.c.max_affinity) \
+            .order_by(CirculationEvent.start.asc())
+
+        if start:
+            query = query.filter(CirculationEvent.start > start)
+
+        if end:
+            query = query.filter(CirculationEvent.end < end)
+
+        results = query.all()
```

Comment:
Now that I'm looking at the whole system, I'm uncomfortable with running this kind of open-ended query on the same machine that is trying to serve books to patrons. Can you send a problem detail instead if the gap between start and end is too wide (the default value of end being right now)? Let's set it at one day to start with.

This way someone who really does want all that data can get it, but they won't be tying up a connection for ten minutes, and they can space out the requests.